File: src/features/typing/data/go/construct_binary_tree_from_preorder_and_inorder_traversal.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	index := make(map[int]int)
	for i, val := range inorder {
		index[val] = i
	}
	var build func(preLeft, preRight, inLeft, inRight int) *TreeNode
	build = func(preLeft, preRight, inLeft, inRight int) *TreeNode {
		if preLeft > preRight {
			return nil
		}
		rootVal := preorder[preLeft]
		mid := index[rootVal]
		leftSize := mid - inLeft
		return &TreeNode{
			Val:   rootVal,
			Left:  build(preLeft+1, preLeft+leftSize, inLeft, mid-1),
			Right: build(preLeft+leftSize+1, preRight, mid+1, inRight),
		}
	}
	return build(0, len(preorder)-1, 0, len(inorder)-1)
}

File: src/features/typing/data/java/construct_binary_tree_from_preorder_and_inorder_traversal.java
import java.util.HashMap;
import java.util.Map;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    private Map<Integer, Integer> index = new HashMap<>();

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for (int i = 0; i < inorder.length; i++) {
            index.put(inorder[i], i);
        }
        return build(preorder, 0, preorder.length - 1, 0, inorder.length - 1);
    }

    private TreeNode build(int[] preorder, int preLeft, int preRight, int inLeft, int inRight) {
        if (preLeft > preRight) {
            return null;
        }
        int rootVal = preorder[preLeft];
        int mid = index.get(rootVal);
        int leftSize = mid - inLeft;
        TreeNode root = new TreeNode(rootVal);
        root.left = build(preorder, preLeft + 1, preLeft + leftSize, inLeft, mid - 1);
        root.right = build(preorder, preLeft + leftSize + 1, preRight, mid + 1, inRight);
        return root;
    }
}
